// ng2-firebase/src/firebase.service.js
System.register(['angular2/core', "./firebase-utils", "./firebase-array"], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var core_1, firebase_utils_1, firebase_array_1;
    var FirebaseService;
    return {
        setters:[
            function (core_1_1) {
                core_1 = core_1_1;
            },
            function (firebase_utils_1_1) {
                firebase_utils_1 = firebase_utils_1_1;
            },
            function (firebase_array_1_1) {
                firebase_array_1 = firebase_array_1_1;
            }],
        execute: function() {
            /**
             * Defines a service that wraps the Firebase Javascript API in a nice, Observable-enabled manner.
             *
             * **Example**:
             *
             * ```TypeScript
             * import {FirebaseService} from 'ng2-firebase/core';
             *
             * // Tell TypeScript that the Firebase SDK has created a global for us
             * declare var Firebase;
             *
             * var firebase = new FirebaseService(
             *    new Firebase('https://YOUR-FIREBASE-URL.firebaseio-demo.com')
             * );
             *
             * // Use Service
             * ```
             *
             * **Angular 2 Example:**
             * ```
             * // some.component.ts
             * import { Component, OnInit, provide } from 'angular2/core';
             * import { FirebaseService, FirebaseServiceFactory } from 'ng2-firebase/core';
             * import { Observable } from 'rxjs/Rx';
             *
             * @@Component({
             *    // FirebaseServiceFactory is not Implemented yet...
             *    selector: 'some-component',
             *
             *    // Make sure to include the async pipe so that the most recent value
             *    // is resolved from the data observable.
             *    template: 'My Data: {{data | async}}',
             *
             *    // Declare the providers that should be used for the service.
             *    providers: [
             *      provide(FirebaseService, { useFactory: FirebaseServiceFactory })
             *    ]
             * })
             * export class SomeComponent implements OnInit {
             *   private firebase: FirebaseService;
             *   data: Observable<any>;
             *
             *   constructor(firebase: FirebaseService) {
             *      this.firebase = firebase;
             *   }
             *
             *   observeData() {
             *      this.data = this.firebase.data;
             *   }
             *
             *   ngOnInit() {
             *      this.observeData();
             *   }
             * }
             * ```
             */
            FirebaseService = (function () {
                /**
                 * Creates a new FirebaseService using the given Firebase JavaScript API Object.
                 * @param firebase The Object that represents the instantiated Firebase JavaScript API Object.
                 */
                function FirebaseService(firebase) {
                    this._firebase = firebase;
                }
                Object.defineProperty(FirebaseService.prototype, "firebase", {
                    /**
                     * Gets the internal Firebase Instance.
                     * @returns {Firebase}
                     */
                    get: function () {
                        return this._firebase;
                    },
                    enumerable: true,
                    configurable: true
                });
                /**
                 * Wraps the given Firebase event type as an observable.
                 * @param eventType {string} One of the following strings: "value", "child_added", "child_changed", "child_removed", or "child_moved."
                 */
                FirebaseService.prototype.wrapFirebaseEvent = function (eventType) {
                    return firebase_utils_1.FirebaseUtils.wrapFirebaseEvent(this.firebase, eventType);
                };
                /**
                 * Retrieves an observable that wraps the given event from the Firebase API.
                 * @param eventType {string} One of the following strings: "value", "child_added", "child_changed", "child_removed", or "child_moved."
                 * @returns {Observable<FirebaseDataSnapshot>} An object that represents the asynchronous stream of events.
                 */
                FirebaseService.prototype.on = function (eventType) {
                    return this.wrapFirebaseEvent(eventType);
                };
                Object.defineProperty(FirebaseService.prototype, "valueRaw", {
                    /**
                     * Gets the raw event stream for the 'value' event from the underlying Firebase Object.
                     * @returns {Observable<any>}
                     */
                    get: function () {
                        return this.wrapFirebaseEvent('value');
                    },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(FirebaseService.prototype, "value", {
                    /**
                     * Gets an observable that resolves with the value in this Firebase location and whenever the data is updated.
                     * Internally, this maps to the 'value' event emitted by Firebase.
                     * @returns {Observable<any>}
                     */
                    get: function () {
                        return this.valueRaw.map(function (data) { return data[0].val(); });
                    },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(FirebaseService.prototype, "dataRaw", {
                    /**
                     * Alias for .valueRaw.
                     * @returns {Observable<any>}
                     */
                    get: function () {
                        return this.valueRaw;
                    },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(FirebaseService.prototype, "data", {
                    /**
                     * Gets an observable that resolves with the data in this Firebase location and whenever the data is updated.
                     * Semantically the same as calling .value.
                     * Internally, this maps to the 'value' event emitted by Firebase.
                     * @returns {Observable<any>}
                     */
                    get: function () {
                        return this.value;
                    },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(FirebaseService.prototype, "childAddedRaw", {
                    /**
                     * Gets the raw event stream for the 'child_added' event from the underlying Firebase Object.
                     * @returns {Observable<any>}
                     */
                    get: function () {
                        return this.wrapFirebaseEvent('child_added');
                    },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(FirebaseService.prototype, "childAdded", {
                    /**
                     * Gets an observable that resolves whenever a child is added to this Firebase location.
                     * Internally, this maps to the 'child_added' event emitted by Firebase.
                     * @returns {Observable<any>}
                     */
                    get: function () {
                        return this.childAddedRaw.map(function (data) { return data[0].val(); });
                    },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(FirebaseService.prototype, "childChangedRaw", {
                    /**
                     * Gets the raw event stream for the 'child_changed' event from the underlying Firebase Object.
                     * @returns {Observable<any>}
                     */
                    get: function () {
                        return this.wrapFirebaseEvent('child_changed');
                    },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(FirebaseService.prototype, "childChanged", {
                    /**
                     * Gets an observable that resolves whenever the data of a child in this Firebase location is modified.
                     * Internally, this maps to the 'child_changed' event emitted by Firebase.
                     * @returns {Observable<any>}
                     */
                    get: function () {
                        return this.childChangedRaw.map(function (data) { return data[0].val(); });
                    },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(FirebaseService.prototype, "childRemovedRaw", {
                    /**
                     * Gets the raw event stream for the 'child_removed' event from the underlying Firebase Object.
                     * @returns {Observable<any>}
                     */
                    get: function () {
                        return this.wrapFirebaseEvent('child_removed');
                    },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(FirebaseService.prototype, "childRemoved", {
                    /**
                     * Gets an observable that resolves whenever a child is removed from this Firebase location.
                     * Internally, this maps to the 'child_removed' event emitted by Firebase.
                     * @returns {Observable<any>}
                     */
                    get: function () {
                        return this.childRemovedRaw.map(function (data) { return data[0].val(); });
                    },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(FirebaseService.prototype, "childMoved", {
                    /**
                     * Gets an observable that resolves whenever a child is moved in this Firebase location.
                     * Internally, this maps to the 'child_moved' event emitted by Firebase.
                     * @returns {Observable<any>}
                     */
                    get: function () {
                        return this.childMovedRaw.map(function (data) { return data[0].val(); });
                    },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(FirebaseService.prototype, "childMovedRaw", {
                    /**
                     * Gets the raw event stream for the 'child_moved' event from the underlying Firebase Object.
                     * @returns {Observable<any>}
                     */
                    get: function () {
                        return this.wrapFirebaseEvent('child_moved');
                    },
                    enumerable: true,
                    configurable: true
                });
                /**
                 * Sets the data exact data that this Firebase location should contain and
                 * returns an observable that represents the operation.
                 * @param data The data that should be set to this location.
                 * @returns {Promise<boolean>} Returns a promise that resolves `true` if the data was set. Otherwise the promise rejects if there was an error.
                 */
                FirebaseService.prototype.setData = function (data) {
                    return firebase_utils_1.FirebaseUtils.wrapFirebaseAsyncCall(this.firebase, this.firebase.set, [data]).then(function () { return true; });
                };
                /**
                 * @alias setData(data)
                 */
                FirebaseService.prototype.set = function (data) {
                    return this.setData(data);
                };
                /**
                 * Update the objects children in this Firebase location. Passing null to updateData() will remove the value at the specified location.
                 * returns an promise that resolves when the operation is succesfull.
                 * @param data The object containing only the keys that should be updated in this location.
                 * @returns {Promise<boolean>} Returns a promise that resolves `true` if the data was succesfully updated. Otherwise the promise rejects if there was an error.
                 */
                FirebaseService.prototype.updateData = function (data) {
                    return firebase_utils_1.FirebaseUtils.wrapFirebaseAsyncCall(this.firebase, this.firebase.update, [data]).then(function () { return true; });
                };
                /**
                 * @alias updateData(data)
                 */
                FirebaseService.prototype.update = function (data) {
                    return this.updateData(data);
                };
                /**
                 * Adds the given data to this Firebase location.
                 * @param data The data that should be added.
                 * @returns {Promise<boolean>}
                 */
                FirebaseService.prototype.push = function (data) {
                    return firebase_utils_1.FirebaseUtils.wrapFirebaseAsyncCall(this.firebase, this.firebase.push, [data]).then(function () { return true; });
                };
                /**
                 * Removes the child with the given key from this location.
                 * If a key is not provided, then this location will be removed from it's parent location.
                 * @param key The key of the child that should be removed from this location.
                 * @returns {Promise<boolean>}
                 */
                FirebaseService.prototype.remove = function (key) {
                    var firebase = this.firebase;
                    if (key) {
                        firebase = firebase.child(key);
                    }
                    return firebase_utils_1.FirebaseUtils.wrapFirebaseAsyncCall(firebase, this.firebase.remove, []).then(function () { return true; });
                };
                /**
                 * Wraps this FirebaseService in a new FirebaseArray object.
                 * The FirebaseArray service provides functionality for dealing with synchronized order lists of objects.
                 * @returns {FirebaseArray}
                 */
                FirebaseService.prototype.asArray = function () {
                    return new firebase_array_1.FirebaseArray(this);
                };
                /**
                 * Gets a child Firebase service that represents the data at the given path.
                 * @param path The relative path from this Firebase location to the requested location.
                 * @returns {FirebaseService}
                 */
                FirebaseService.prototype.child = function (path) {
                    return new FirebaseService(this.firebase.child(path));
                };
                FirebaseService = __decorate([
                    core_1.Injectable(), 
                    __metadata('design:paramtypes', [Object])
                ], FirebaseService);
                return FirebaseService;
            }());
            exports_1("FirebaseService", FirebaseService);
        }
    }
});
